sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,295 @@
+		var revisit []*Item
+		deleted, revisit, err = s.identifyItemsAbsentOrShadowed(bucketKey, maybeDeleted)
+		if err != nil {
+			// Return partial information?
+			return present, nil, err
+		}
+		present = append(present, revisit...)
+	}
+
+	// Swap the replacement bucket in to the cache (or delete)
+	cacheKey := s.GetCacheKey(bucketKey)
+	s.bucketsCacheLock.Lock()
+	if bucketRemoved {
+		s.bucketsCache.Delete(cacheKey)
+	} else {
+		s.bucketsCache.Insert(cacheKey, replacementBucket)
+	}
+	s.bucketsCacheLock.Unlock()
+
+	return present, deleted, nil
+
+}
+
+// Look up each item and determine whether another copy (or version) of it exists
+// in a different bucket.
+func (s *StoragePackerV2) identifyItemsAbsentOrShadowed(bucketKey string, maybeDeleted []*itemRequest) ([]*Item, []*Item, error) {
+	deleted := make([]*Item, 0)
+	revisit := make([]*Item, 0)
+
+	// Using the radix lookup will find child buckets, but not parent
+	// buckets.  For that we'll have to walk up the tree a step at a time.
+	//
+	// For example, we might be processing 00/0
+	//
+	// 00 -- 00/0 -- 00/0/a
+	//
+	// but, the item with key 000a... might exist in 00/0/a or maybe even in 00,
+	// so we want to check both.
+	//
+	// If we find it somewhere, return the version in the bucket with the longest key.
+	// (That is more likely to be a more recent version.)
+	partition, err := s.partitionRequests(maybeDeleted)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem searching for deleted keys: {{err}}", err)
+	}
+
+	// Looking through the other buckets without acquiring a lock
+	// is unsafe, but should be OK in the context of a replica.
+	for _, p := range partition {
+		bucketsToCheck := s.bucketAndAllParents(p.Bucket)
+		for _, request := range p.Requests {
+			found := false
+			for _, b := range bucketsToCheck {
+				// Skip originalBucket
+				if b.Key != bucketKey {
+					var data []byte
+					if data, found = b.ItemMap[request.ID]; found {
+						revisit = append(revisit, &Item{
+							ID:   request.ID,
+							Data: data,
+						})
+						break
+					}
+				}
+			}
+			if !found {
+				// Include the original value
+				// though maybe this is overkill and we should just return IDs.
+				deleted = append(deleted, request.Value)
+			}
+		}
+	}
+
+	return deleted, revisit, nil
+
+}
+
+// Assemble a list (from longest key to shortest) of all the parent buckets.
+func (s *StoragePackerV2) bucketAndAllParents(bucket *LockedBucket) []*LockedBucket {
+	s.bucketsCacheLock.RLock()
+	defer s.bucketsCacheLock.RUnlock()
+
+	parents := []*LockedBucket{bucket}
+	for k := s.parentCacheKey(bucket.Key); k != ""; k = s.parentCacheKey(k) {
+		bucketRaw, found := s.bucketsCache.Get(k)
+		if found {
+			parents = append(parents, bucketRaw.(*LockedBucket))
+		}
+	}
+	return parents
+}
+
+// Given a bucket key, return the cache key for its parent
+func (s *StoragePackerV2) parentCacheKey(bucketKey string) string {
```

Comment:
What happens if this is called against the topmost bucket?

Reply:
It returns "".